librad/src/git/trailer.rs

```diff
@@ -0,0 +1,273 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+pub enum Error {
+    #[error("The given input is not a well-formed trailer")]
+    IllFormed,
+
+    #[error("Could not parse the trailer due to an invalid token found")]
+    Token(#[from] InvalidToken),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer<'a> {
+    token: Token<'a>,
+    value: &'a str,
+}
+
+impl<'a> Trailer<'a> {
+    #[allow(dead_code)]
+    /// Try and parse a [Trailer] form a given raw &str that is
+    /// expected to represent a trailer in raw form.
+    pub fn parse(s: &'a str, separators: &'a [char]) -> Result<Trailer<'a>, Error> {
+        let parts: Vec<&str> = s.split(&separators[..]).map(|s| s.trim()).collect();
+        match parts.as_slice() {
+            [token, value] => Self::new(*token, *value),
+            _ => Err(Error::IllFormed),
+        }
+    }
+
+    pub fn new(token: &'a str, value: &'a str) -> Result<Trailer<'a>, Error> {
+        Ok(Trailer {
+            token: token.try_into()?,
+            value,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Token<'a>(&'a str);
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+#[error("Invalid token")]
+pub struct InvalidToken;
+
+impl<'a> TryFrom<&'a str> for Token<'a> {
+    type Error = InvalidToken;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        if value.chars().all(|c| c.is_alphanumeric() || c == '-') {
+            Ok(Self(value))
+        } else {
+            Err(InvalidToken)
+        }
+    }
+}
+
+/// Parse the trailers of the given message. It looks up the last paragraph
+/// of the message and attempts to parse each of its lines as a [Trailer].
+/// Fails if at least one trailer fails to be parsed.
+// Todo(nuno):
+//   * return iterator instead of Vec?
+//   * rsplit("\n\n").next() is always Some(_), even for "", so consider simpler
+//     return type.
+pub fn parse<'a>(
+    message: &'a str,
+    separators: &'a [char],
+) -> Result<Option<Vec<Trailer<'a>>>, Error> {
```

Comment:
Ya so isn't `None` the same as `vec![]`?

Reply:
Yeah, they are semantically the same. My goal really was to differentiate between the case where there is no trailers paragraph vs the case where there is a trailers paragraph but they ill-formed. The thing is that for the first case we can't really tell. Shall we just have `Result<Vec<Trailer, Error>>` then? It seems sensible to me.

Reply:
Ya, I think `Result<Vec<Trailer>, Error>` makes more sense to me :)